### PR TITLE
change MDP class to gym wrappers

### PR DESCRIPTION
**Pull request by zhouzypaul**
- remove the MDP & State classes 
- add gym wrappers for goal conditioned MDPs and AntMaze
- adjust the original agent code so it works with the new gym wrappers
- add script for plotting learning curves 
- move all logging data to ./results
- add requirements.txt and .gitignore

### CODE REVIEW

File: hrl/__main__.py [hunk 1]
```diff
@@ -44,12 +54,15 @@
     if args.clear_option_buffers:
         assert not args.use_global_value_function
 
-    if args.environment == "umaze":
-        mdp = D4RLAntMazeMDP("umaze", goal_state=np.array((0, 8)), seed=args.seed)
+    if args.environment in ["antmaze-umaze-v0", "antmaze-medium-play-v0", "antmaze-large-play-v0"]:
+        env = gym.make(args.environment)
+        env = D4RLAntMazeWrapper(env, start_state=((0, 0)), goal_state=np.array((0, 8)), use_dense_reward=args.use_dense_rewards)
```
Comment (abagaria) [inline hrl/__main__.py]: The goal state for umaze is at (0, 8); that is not true for the other mazes.

File: hrl/__main__.py [hunk 2]
```diff
@@ -44,12 +54,15 @@
...
+        seeding.seed(0, random, torch, np)
```
Comment (abagaria) [inline hrl/__main__.py:70]: Can't we do `seeding.seed(args.seed, gym, env, random, torch, np)`? Why seed some with 0 and others with `args.seed`?

File: hrl/agent/dsc/MBOptionClass.py [hunk 3]
```diff
@@ -221,32 +222,33 @@ def rollout(self, step_number, rollout_goal=None, eval_mode=False):
...
 
         self.num_executions += 1
 
-        while not self.is_at_local_goal(state, goal) and step_number < self.max_steps and num_steps < self.timeout:
+        while not self.is_at_local_goal(state, done, goal) and step_number < self.max_steps and num_steps < self.timeout:
 
             # Control
             action = self.act(state, goal)
-            reward, next_state = self.mdp.execute_agent_action(action)
+            next_state, reward, next_done, _ = self.mdp.step(action)
```
Comment (abagaria) [inline hrl/agent/dsc/MBOptionClass.py:228]: Worth checking: `next_done` should always be for "real" terminal states and not timeouts, which are common in gym.

File: hrl/agent/dsc/dsc.py [hunk 4]
```diff
@@ -62,32 +59,34 @@ def _pick_earliest_option(state, options):
             if option.is_init_true(state) and not option.is_term_true(state):
                 return option
 
-    def act(self, state):
+    def act(self, state, done):
```
Comment (abagaria) [inline hrl/agent/dsc/dsc.py]: It feels kinda weird to have a `done` flag as in input to an `act()` function. Can we make `state` the only input to this function?

File: hrl/agent/dsc/utils.py [hunk 5]
```diff
@@ -114,7 +116,8 @@ def plot_two_class_classifier(option, episode, experiment_name, plot_examples=Tr
 
     name = option.name if episode is None else option.name + "_{}_{}".format(experiment_name, episode)
     plt.title("{} Initiation Set".format(option.name))
-    plt.savefig("initiation_set_plots/{}/{}_initiation_classifier_{}.png".format(experiment_name, name, seed))
+    saving_path = os.path.join('results', experiment_name, 'initiation_set_plots', '{}_initiation_classifier_{}.png'.format(name, seed))
```
Comment (abagaria) [inline hrl/agent/dsc/utils.py]: `f"{name}_initiation_classifier_{seed}.png"` is less verbose and more modern

File: hrl/wrappers/antmaze_wrapper.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+import numpy as np
+import torch
+
+from hrl.wrappers.gc_mdp_wrapper import GoalConditionedMDPWrapper
+
+
+class D4RLAntMazeWrapper(GoalConditionedMDPWrapper):
+	def __init__(self, env, start_state, goal_state, use_dense_reward=False):
+		self.env = env
+		self.norm_func = lambda x: np.linalg.norm(x, axis=-1) if isinstance(x, np.ndarray) else torch.norm(x, dim=-1)
```
Comment (abagaria) [inline hrl/wrappers/antmaze_wrapper.py:12]: Nice, I like this distance function.

File: hrl/agent/dsc/MBOptionClass.py [hunk 7]
```diff
@@ -128,8 +128,9 @@ def get_training_phase(self):
         return "initiation_done"
 
     def extract_features_for_initiation_classifier(self, state):
```
Comment (abagaria) [inline hrl/agent/dsc/MBOptionClass.py]: Looks like we have this function both in the MBOptionClass and the GoalDirectedEnvWrapper. Can we limit to defining it in one place?

File: hrl/wrappers/gc_mdp_wrapper.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+        self.init_state = self.env.reset()
+        self.cur_state = deepcopy(self.init_state)
+        self.cur_done = False
+        return self.init_state
+    
+    def step(self, action):
+        """
+        overwrite the step function for gc MDP.
+        """
+        next_state, reward, done, info = self.env.step(action)
```
Comment (abagaria) [inline hrl/wrappers/gc_mdp_wrapper.py:58]: I am worried that `done` could be true for time outs. Can we prevent that and limit `done` to be `True` only for actual termination states?

File: hrl/wrappers/antmaze_wrapper.py [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+	
+	def sparse_gc_reward_func(self, states, goals, batched=False):
+		"""
+		overwritting sparse gc reward function for antmaze
+		"""
+		# assert input is np array or torch tensor
+		assert isinstance(states, (np.ndarray, torch.Tensor))
+		assert isinstance(goals, (np.ndarray, torch.Tensor))
+
+		# TODO: change this to batched operation
```
Comment (abagaria) [inline hrl/wrappers/antmaze_wrapper.py]: Delete TODO, I think it has been done :)

**Review by abagaria — CHANGES_REQUESTED** [review]
Looks pretty good, suggesting some minor changes. In the future, I want to see if we can get rid of `cur_state` and `cur_done`, but lets talk about that some other time.

File: hrl/__main__.py (hunk 1 above)
Comment (zhouzypaul) [reply hrl/__main__.py]: what are the goal state for other mazes?

Comment (zhouzypaul) [reply hrl/__main__.py]: I'm thinking maybe we can convert `goal_state` into a command-line argument that needs to be specified.

File: hrl/__main__.py (hunk 2 above)
Comment (zhouzypaul) [reply hrl/__main__.py:70]: I think we can. 

In Cam's codebase, he used it this way -- only using the user-specified seed for the gym environment. I'm not sure what's the reasoning behind it. I think the reason is that when plotting with different random seeds, the neural network is still initialized the same way because the torch seed is the same, and so the randomness only comes from the gym environment.

File: hrl/agent/dsc/dsc.py (hunk 4 above)
Comment (zhouzypaul) [reply hrl/agent/dsc/dsc.py]: I agree. 

But the problem is that the `is_at_local_goal()` function, which is called inside `act()`, needs `done` as an argument.

File: hrl/agent/dsc/utils.py (hunk 5 above)
Comment (zhouzypaul) [reply hrl/agent/dsc/utils.py]: I intensionally used `.format()` to extend support for Python2, which doesn't support formatted strings. 

But if we don't need to consider support for Python2, I'm more than happy to change it to formatted strings.

File: hrl/agent/dsc/MBOptionClass.py (hunk 7 above)
Comment (zhouzypaul) [reply hrl/agent/dsc/MBOptionClass.py]: good point!

Comment (zhouzypaul) [reply hrl/agent/dsc/MBOptionClass.py]:
```python
if "push" in self.mdp.unwrapped.spec.id:
            return features[:4]
```
is this ever used?

File: hrl/wrappers/antmaze_wrapper.py (hunk 9 above)
Comment (zhouzypaul) [reply hrl/wrappers/antmaze_wrapper.py]: Ah, I forgot about this one. 

I was actually hoping (at some point) we could remove the `batched` argument in this function, and make sure that all input `states` and `goals` are batched.

File: hrl/__main__.py (hunk 1 above)
Comment (abagaria) [reply hrl/__main__.py]: Ok that sounds good.

Comment (abagaria) [reply hrl/__main__.py]: If no goal is provided, can we ask d4rl for a goal state?

-

File: hrl/agent/dsc/utils.py (hunk 5 above)
Comment (abagaria) [reply hrl/agent/dsc/utils.py]: Yea, we don't need to support py2

File: hrl/agent/dsc/MBOptionClass.py (hunk 7 above)
Comment (abagaria) [reply hrl/agent/dsc/MBOptionClass.py]: Rather than having env specific code in the Option class, I would rather push that responsibility to the Env

File: hrl/wrappers/antmaze_wrapper.py (hunk 9 above)
Comment (abagaria) [reply hrl/wrappers/antmaze_wrapper.py]: Agreed! Everything should always be batched and a single state should have dim (1, D) rather than (D,)

File: hrl/agent/dsc/MBOptionClass.py [hunk 10]
```diff
@@ -157,11 +158,11 @@ def pessimistic_is_init_true(self, state):
...
         return self.pessimistic_classifier.predict([features])[0] == 1
 
-    def is_at_local_goal(self, state, goal):
+    def is_at_local_goal(self, state, done, goal):
         """ Goal-conditioned termination condition. """
 
-        reached_goal = self.mdp.sparse_gc_reward_function(state, goal, {})[1]
-        reached_term = self.is_term_true(state) or state.is_terminal()
+        reached_goal = self.mdp.sparse_gc_reward_func(state, goal)[1]
+        reached_term = self.is_term_true(state) or done
```
Comment (abagaria) [inline hrl/agent/dsc/MBOptionClass.py]: Come to think of it, I don't see why `reached_term` is true for terminal states. If the terminal state is a goal in the MDP, then `reached_goal` will anyway be true. If it is a "death" state, `reached_term` should anyway **not** be true. So, can we get rid of the `done` flag in `is_at_local_goal()`? That way, we will also not need `done` in the `act()` function.

**Review by abagaria — APPROVED** [review]
Pulling @pzhou's changes to get rid of our reliance on MDPClass and StateClass

File: hrl/agent/dsc/MBOptionClass.py (hunk 10 above)
Comment (zhouzypaul) [reply hrl/agent/dsc/MBOptionClass.py]: Yea I believe that's already done at #91ad5b2, the code for your comment was outdated
